flow/aggregates.py

```diff
@@ -0,0 +1,481 @@
+        if not self._is_aggregate:
+            return _DefaultAggregateStore(project)
+        else:
+            return _AggregatesStore(self, project)
+
+    def __call__(self, func=None):
+        if callable(func):
+            setattr(func, '_flow_aggregate', self)
+            return func
+        else:
+            raise TypeError('Invalid argument passed while calling '
+                            'the aggregate instance. Expected a callable, '
+                            f'got {type(func)}.')
+
+
+class _AggregatesStore:
+    """This class holds the information of all the aggregates associated with
+    a :py:class:`aggregator`.
+
+    This is a callable class which, when called, generates all the aggregates.
+    When iterated through it's instance, all the aggregates are yielded.
+
+    :param aggregate:
+        aggregator object associated with this class.
+    :type aggregate:
+        :py:class:`aggregator`
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, aggregate, project):
+        self._aggregate = aggregate
+
+        # We need to register the aggregates for this instance using the
+        # project provided.
+        self._aggregates = list()
+        self._aggregate_ids = dict()
+        self._register_aggregates(project)
+
+    def __iter__(self):
+        yield from self._aggregates
+
+    def __getitem__(self, id):
+        "Return an aggregate, if exists, using the id provided"
+        try:
+            return self._aggregate_ids[id]
+        except KeyError:
+            raise LookupError(f'Unable to find the aggregate having id {id} in '
+                              'the FlowProject')
+
+    def __contains__(self, aggregate):
+        """Return whether an aggregate is stored in the this
+        instance of :py:class:`_AggregateStore`
+
+        :param aggregate:
+            An aggregate of jobs.
+        :type aggregate:
+            tuple of :py:class:`signac.contrib.job.Job`
+        """
+        return get_aggregate_id(aggregate) in self._aggregate_ids
+
+    def __len__(self):
+        return len(self._aggregates)
+
+    def __eq__(self, other):
+        return type(self) == type(other) and \
+               self._aggregates == other._aggregates and \
+               self._aggregate == other._aggregate
+
+    def __hash__(self):
+        blob = str(hash(self._aggregate))
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _register_aggregates(self, project):
+        """If the instance of this class is called then we will
+        generate aggregates and store them in ``self._aggregates``.
+        """
+        make_aggregates = _MakeAggregates(self._aggregate, project)
+        aggregates, aggregate_ids = make_aggregates()
+        self._aggregates = aggregates
+        self._aggregate_ids = aggregate_ids
```

Comment:
If we are going to use `_MakeAggregates` then we would just want this class to have a `_MakeAggregates` object it can call when it needs to.